SYCL/ESIMD/api/slm_gather_scatter_rgba.cpp

```diff
@@ -0,0 +1,187 @@
+  case rgba_channel_mask::ABGR:
+    return "ABGR";
+  default:
+    return "";
+  }
+  return "";
+}
+
+template <typename T, unsigned VL, auto CH_MASK> bool test(queue q) {
+  using namespace sycl::ext::intel::experimental::esimd;
+  constexpr int numChannels = get_num_channels_enabled(CH_MASK);
+  constexpr size_t size = VL * numChannels;
+
+  std::cout << "Testing T=" << typeid(T).name() << " VL=" << VL
+            << " MASK=" << convertMaskToStr(CH_MASK) << "...\n";
+
+  auto dev = q.get_device();
+  auto ctxt = q.get_context();
+  T *A = static_cast<T *>(malloc_shared(size * sizeof(T), dev, ctxt));
+  T *gold = new T[size];
+
+  for (int i = 0; i < size; ++i) {
+    A[i] = (T)-i;
+  }
+
+  // Fill out the array with gold values.
+  // R  R  R  R   ... G  G  G  G   ... B  B  B   B   ... A  A  A   A  ...
+  // 0, 4, 8, 12, ... 1, 5, 9, 13, ... 2, 6, 10, 14, ... 3, 7, 11, 15 ...
+  for (unsigned i = 0; i < numChannels; i++)
+    for (unsigned j = 0; j < VL; j++)
+      gold[i * VL + j] = j * numChannels + i;
+
+  // Account for masked out last lanes (with pred argument to slm_gather_rgba).
+  unsigned maskedIndex = VL - 1;
+  for (unsigned i = 0; i < numChannels; i++, maskedIndex += VL)
+    gold[maskedIndex] = 0;
+
+  try {
+    // We need that many workitems
+    sycl::range<1> GlobalRange{1};
+    // Number of workitems in a workgroup
+    sycl::range<1> LocalRange{1};
+    sycl::nd_range<1> Range{GlobalRange * LocalRange, LocalRange};
+
+    auto e = q.submit([&](handler &cgh) {
+      Kernel<T, VL, CH_MASK> kernel(A);
+      cgh.parallel_for(Range, kernel);
+    });
+    e.wait();
+  } catch (sycl::exception const &e) {
+    std::cerr << "SYCL exception caught: " << e.what() << '\n';
+    free(A, ctxt);
+    delete[] gold;
+    return static_cast<bool>(e.code());
+  }
+
+  int err_cnt = 0;
+  for (unsigned i = 0; i < size; ++i) {
+    if (A[i] != gold[i]) {
+      if (++err_cnt < 35) {
+        std::cerr << "failed at index " << i << ": " << A[i]
+                  << " != " << gold[i] << " (gold)\n";
+      }
+    }
+  }
+
+  if (err_cnt > 0) {
+    std::cout << "  pass rate: "
+              << ((float)(size - err_cnt) / (float)size) * 100.0f << "% ("
+              << (size - err_cnt) << "/" << size << ")\n";
+  }
+
+  free(A, ctxt);
+  delete[] gold;
+
+  std::cout << (err_cnt > 0 ? "  FAILED\n" : "  Passed\n");
+  return err_cnt > 0 ? false : true;
+}
+
+template <typename T, unsigned VL> bool test(queue q) {
+  using namespace sycl::ext::intel::experimental::esimd;
+  bool passed = true;
+  passed &= test<T, VL, rgba_channel_mask::R>(q);
+  passed &= test<T, VL, rgba_channel_mask::GR>(q);
+  passed &= test<T, VL, rgba_channel_mask::ABGR>(q);
+  return passed;
+}
+
+int main(void) {
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+  bool passed = true;
+  passed &= test<int, 8>(q);
```

Comment:
I'd suggest to add char and uint16_t cases, these are usually the most problematic cases.

Reply:
If they fail - quick solution would be JIRA + TODO in the test

Reply:
I did not find these intrinsics in the ESIMD DOC: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md 

But the implementation exists only for 4-byte elements: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp#L698
https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/experimental/esimd/memory.hpp#L730

Reply:
OK, looks like we need JIRAs on this.
Approving this one.